### ISSUE TITLE
Windows build linking issues with hdf5

### ISSUE THREAD

**Issue by dagewa** [cctbx/dxtbx#329]
On [Windows](https://github.com/dials/dials/wiki/DIALS-on-Windows) there are problems with NeXus. Various tests fail within this call:
https://github.com/cctbx/dxtbx/blob/36b576b077ebdcf717773ea1fe13ef769498cd22/format/nexus.py#L59
On the C++ side, that ends up here:
https://github.com/cctbx/dxtbx/blob/36b576b077ebdcf717773ea1fe13ef769498cd22/format/boost_python/nexus_ext.cc#L75-L78
This fails someway down the line, but I can immediately see that the `dataset_id` here is not valid by adding these lines at the top of the function:
```
    htri_t obj_valid = H5Iis_valid( dataset_id );
    std::cout << "obj_valid is : " << obj_valid << std::endl;
```
I'm beginning to think this is a linking problem. On Linux `ldd build/lib/dxtbx_format_nexus_ext.so` contains, amongst other things `	libhdf5.so.103 => /home/fcx32934/sw/cctbx/conda_base/lib/libhdf5.so.103 (0x00007fe1f9e7a000)`. However, on Windows I see this:

```
c:\Users\fcx32934\sw\cctbx-conda-compilers>dumpbin /dependents build\lib\dxtbx_format_nexus_ext.pyd
Microsoft (R) COFF/PE Dumper Version 14.28.29336.0
Copyright (C) Microsoft Corporation.  All rights reserved.


Dump of file build\lib\dxtbx_format_nexus_ext.pyd

File Type: DLL

  Image has the following dependencies:

    python38.dll
    boost_python38.dll
    zlib.dll
    MSVCP140.dll
    KERNEL32.dll
    VCRUNTIME140.dll
    api-ms-win-crt-runtime-l1-1-0.dll
    api-ms-win-crt-heap-l1-1-0.dll
    api-ms-win-crt-environment-l1-1-0.dll
    api-ms-win-crt-string-l1-1-0.dll
    api-ms-win-crt-stdio-l1-1-0.dll
    api-ms-win-crt-math-l1-1-0.dll
    api-ms-win-crt-convert-l1-1-0.dll
    api-ms-win-crt-filesystem-l1-1-0.dll
    api-ms-win-crt-time-l1-1-0.dll
    api-ms-win-crt-utility-l1-1-0.dll

  Summary

        A000 .data
       11000 .pdata
       69000 .rdata
        1000 .reloc
        1000 .rsrc
      1AC000 .text
        8000 _RDATA
```
No hint of a link to anything `hdf5`.

Does anybody have any suggestions on how to explore further?

**Comment by Anthchirp** [comment]
The file to look at is definitely `SConscript`, probably around
https://github.com/cctbx/dxtbx/blob/36b576b077ebdcf717773ea1fe13ef769498cd22/SConscript#L39

In particular, there are multiple suspicious references to specific HDF5 1.8 versions in specific places, which indicates to me that a lot of stuff in there is not generic, fragile, and likely out of date.

**Comment by dagewa** [comment]
Thanks, sounds "fun"

**Comment by dagewa** [comment]
I'm making some progress in https://github.com/cctbx/dxtbx/tree/win-nexus. One sticking point seems to be whether `dxtbx_hdf5_libs` should include `"hdf5"` like on Mac and Linux, or `"libhdf5"`, as it is at present for Windows. It looks like the former is a dynamic library while the latter is static. Perhaps the NeXus problems going from the Python to the C++ side are because h5py is using the dynamic library, while we link to the static? I'm not sure.

I tried changing this to `"hdf5"` but then the build fails as if it can't find `hdf5.h`, even though I corrected the include path. However, I've just found that to use the shared libraries on Windows we need a compile definition `H5_BUILT_AS_DYNAMIC_LIB` to be set.

More next week.
